File: native/android/app/src/main/java/com/example/hybrid/bridge/plugins/CameraPlugin.kt
package com.example.hybrid.bridge.plugins

import android.Manifest
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import com.example.hybrid.bridge.BridgePlugin
import com.example.hybrid.bridge.BridgeResult
import org.json.JSONObject

class CameraPlugin(private val context: android.content.Context) : BridgePlugin {
    override val namespace: String = "device.camera"

    override fun handle(method: String, params: JSONObject, completion: (BridgeResult) -> Unit) {
        when (method) {
            "takePhoto" -> takePhoto(params, completion)
            else -> completion(BridgeResult(-1, "method not found: $method", null))
        }
    }

    private fun takePhoto(params: JSONObject, completion: (BridgeResult) -> Unit) {
        val source = params.optString("source", "camera")
        // 1. 权限检查
        val permission = if (source == "album") Manifest.permission.READ_EXTERNAL_STORAGE
        else Manifest.permission.CAMERA

        if (ActivityCompat.checkSelfPermission(context, permission) != PackageManager.PERMISSION_GRANTED) {
            // 实际项目应在 Activity 中请求权限，这里返回未授权
            completion(BridgeResult(-2, "权限未授予", null))
            return
        }
        // 2. 调起相机 / 相册
        // 实际项目接入 CameraX 或 Intent(ACTION_IMAGE_CAPTURE)
        completion(BridgeResult(0, "ok", JSONObject().put("uri", "file:///mock/photo.jpg")))
    }
}

File: native/ios/HybridApp/Bridge/Plugins/CameraPlugin.swift
import AVFoundation
import Photos

/// 相机插件：拍照 / 选图
final class CameraPlugin: BridgePlugin {
    weak var hostVC: UIViewController?
    var bridge: JSBridgeContentController?
    let namespace = "device.camera"

    func handle(method: String, params: Any, completion: @escaping (BridgeResult) -> Void) {
        switch method {
        case "takePhoto":
            takePhoto(params: params, completion: completion)
        default:
            completion(BridgeResult(code: -1, msg: "method not found: \(method)", data: nil))
        }
    }

    private func takePhoto(params: Any, completion: @escaping (BridgeResult) -> Void) {
        let dict = params as? [String: Any] ?? [:]
        let source = dict["source"] as? String ?? "camera"

        if source == "album" {
            pickFromAlbum(completion: completion)
        } else {
            takeFromCamera(completion: completion)
        }
    }

    private func takeFromCamera(completion: @escaping (BridgeResult) -> Void) {
        // 1. 权限检查
        let status = AVCaptureDevice.authorizationStatus(for: .video)
        guard status == .authorized else {
            if status == .notDetermined {
                AVCaptureDevice.requestAccess(for: .video) { granted in
                    DispatchQueue.main.async {
                        if granted { self.presentCamera(completion: completion) }
                        else { completion(BridgeResult(code: -2, msg: "相机权限被拒绝", data: nil)) }
                    }
                }
                return
            }
            completion(BridgeResult(code: -2, msg: "相机权限被拒绝", data: nil))
            return
        }
        presentCamera(completion: completion)
    }

    private func presentCamera(completion: @escaping (BridgeResult) -> Void) {
        // 实际项目接入 UIImagePickerController 或自研相机
        completion(BridgeResult(code: 0, msg: "ok", data: ["uri": "file:///mock/photo.jpg"]))
    }

    private func pickFromAlbum(completion: @escaping (BridgeResult) -> Void) {
        completion(BridgeResult(code: 0, msg: "ok", data: ["uri": "file:///mock/album.jpg"]))
    }
}
